### PR TITLE
Ability to assign custom open source and closed source llms via settings.

### PR DESCRIPTION
**Pull request by cielecki**
When merging I will squash all the commits

### CODE REVIEW

**Comment by cielecki** [comment]
Look at https://github.com/10clouds/aiconsole/issues/241 to see semantics of what is going on.

File: backend/aiconsole/core/gpt/check_key.py [hunk 1]
```diff
@@ -28,12 +36,15 @@ async def check_key(key: str) -> bool:
         return True
 
     client = OpenAI(api_key=key)
-    try:
-        models = client.models.list().data
-    except AuthenticationError:
-        return False
+    # set key
```
Comment (pbaranski9411) [inline backend/aiconsole/core/gpt/check_key.py:39]: I think this comment is redundant

File: backend/aiconsole/core/gpt/check_key.py [hunk 2]
```diff
@@ -28,12 +36,15 @@ async def check_key(key: str) -> bool:
...
+    openai.api_key = key
+    models = client.models.list()  # type: ignore
```
Comment (pbaranski9411) [inline backend/aiconsole/core/gpt/check_key.py:41]: why type ignore?

File: backend/aiconsole/core/gpt/check_key.py [hunk 3]
```diff
@@ -28,12 +36,15 @@ async def check_key(key: str) -> bool:
...
     available_models = [model.id for model in models]
-    needed_models = MODEL_DATA.keys()
+    needed_models = [
+        GPT_MODE_COST_MODEL,
+        GPT_MODE_QUALITY_MODEL,
+        GPT_MODE_SPEED_MODEL,
+    ]
```
Comment (pbaranski9411) [inline backend/aiconsole/core/gpt/check_key.py:47]: this looks like const. Could be defined on the top of this file

File: backend/aiconsole/core/settings/project_settings.py [hunk 4]
```diff
@@ -201,13 +234,52 @@ async def __load(self) -> SettingsData:
...
+            api_key="openai_api_key",
+        )
+
+        gpt_modes["cost"] = GPTModeConfig(
+            name="cost",
+            max_tokens=GPT_MODE_COST_MAX_TOKENS,
+            encoding=GPTEncoding.GPT_4,
+            model=GPT_MODE_COST_MODEL,
+            api_key="openai_api_key",
+        )
```
Comment (pbaranski9411) [inline backend/aiconsole/core/settings/project_settings.py:262]: this could be defined as one dictionary creation

**Review by pbaranski9411 — APPROVED** [review]

File: backend/aiconsole/core/gpt/request.py [hunk 5]
```diff
@@ -21,13 +21,15 @@
...
-from aiconsole.core.gpt.consts import MODEL_DATA, GPTMode, GPTModel
+from aiconsole.core.gpt.consts import SPEED_GPT_MODE, GPTMode
 from aiconsole.core.gpt.token_error import TokenError
 from aiconsole.core.gpt.types import (
     EnforcedFunctionCall,
     GPTRequestMessage,
     GPTRequestTextMessage,
 )
+from aiconsole.core.settings.project_settings import get_aiconsole_settings
+from aiconsole_toolkit.settings import get_settings
```
Comment (udislau) [inline backend/aiconsole/core/gpt/request.py:32]:
```suggestion

```

**Review by udislau — APPROVED** [review]
Approved, Look out for unused code.

**Pull request merged**